pkg/api/search.go

```diff
@@ -66,8 +67,50 @@ func (hs *HTTPServer) Search(c *models.ReqContext) response.Response {
 		return response.Error(500, "Search failed", err)
 	}
 
-	c.TimeRequest(metrics.MApiDashboardSearch)
-	return response.JSON(http.StatusOK, searchQuery.Result)
+	defer c.TimeRequest(metrics.MApiDashboardSearch)
+
+	if !c.QueryBool("accesscontrol") {
+		return response.JSON(http.StatusOK, searchQuery.Result)
+	}
+
+	return hs.searchHitsWithMetadata(c, searchQuery.Result)
+}
+
+func (hs *HTTPServer) searchHitsWithMetadata(c *models.ReqContext, hits models.HitList) response.Response {
+	folderUIDs := make(map[string]bool, 0)
+	dashboardUIDs := make(map[string]bool, 0)
+	parentUIDs := make(map[string]bool, 0)
+
+	for _, hit := range hits {
+		if hit.Type == models.DashHitFolder {
+			folderUIDs[hit.UID] = true
+		} else {
+			dashboardUIDs[hit.UID] = true
+			parentUIDs[hit.FolderUID] = true
+		}
+	}
+
+	folderMeta := hs.getMultiAccessControlMetadata(c, c.OrgId, dashboards.ScopeFoldersPrefix, folderUIDs)
+	parentMeta := hs.getMultiAccessControlMetadata(c, c.OrgId, dashboards.ScopeFoldersPrefix, parentUIDs)
+	dashboardMeta := hs.getMultiAccessControlMetadata(c, c.OrgId, dashboards.ScopeDashboardsPrefix, dashboardUIDs)
+
+	// search hit with access control metadata attached
+	type hitWithMeta struct {
```

Comment:
Why not add accesscontrol metadata to `models.Hit`? (dependency cycle?)

Reply:
Yes models -> accesscontrol -> models

Reply:
Ok. In case you think it's valuable, for other DTOs in the models package, (ex: `OrgUserDTO`, `TeamDTO`, ...) we used:
`AccessControl map[string]bool 'json:"accessControl,omitempty"'`